### PR TITLE
Colibri web sockets

### PR DESCRIPTION
**Pull request by bgrozev**
Supports WebSockets as an alternative transport for COLIBRI messages (i.e. an alternative to WebRTC data channels).

### CODE REVIEW

File: doc/rest.md [hunk 1]
```diff
@@ -13,6 +13,7 @@ The _public_ interface inludes:
 
 * Support for serving static files (e.g. the HTML/js for jitsi-meet)
 * Support for proxying (for e.g. proxying BOSH connections to a prosody instance)
+* A WebSocket API for communicatoin with conference endpoints (```/colibri/ws/```)
```
Comment (bbaldino) [inline doc/rest.md]: typo

File: src/main/java/org/jitsi/videobridge/Conference.java [hunk 2]
```diff
@@ -1312,7 +1312,8 @@ IceUdpTransportManager getTransportManager(
                 try
                 {
                     transportManager
-                        = new IceUdpTransportManager(this, initiator, 1);
+                        = new IceUdpTransportManager(
+                            this, initiator, 1, channelBundleId);
```
Comment (bbaldino) [inline src/main/java/org/jitsi/videobridge/Conference.java]: what was this change for?

File: src/main/java/org/jitsi/videobridge/EndpointMessageTransport.java [hunk 3]
```diff
@@ -103,46 +140,114 @@
...
+     * @param dst the {@link ColibriWebSocket} through which to send the message.
+     * @param message the message to send.
+     * @param errorMessage an error message to be logged in case of failure.
+     */
+    private void sendMessage(
+        ColibriWebSocket dst, String message, String errorMessage)
+    {
+        try
+        {
+            dst.getRemote().sendString(message);
```
Comment (bbaldino) [inline src/main/java/org/jitsi/videobridge/EndpointMessageTransport.java]: should we add a stat for websocket sends?

File: src/main/java/org/jitsi/videobridge/EndpointMessageTransport.java [hunk 4]
```diff
@@ -464,54 +574,128 @@ public void onStringData(WebRtcDataStream src, String msg)
...
+    void sendMessage(String msg)
         throws IOException
     {
         SctpConnection sctpConnection = endpoint.getSctpConnection();
+        ColibriWebSocket webSocket = this.webSocket;
         String endpointId = endpoint.getID();
 
-        if (sctpConnection == null)
+        Object dst = null;
+        if (webSocketLastActive)
```
Comment (bbaldino) [inline src/main/java/org/jitsi/videobridge/EndpointMessageTransport.java]: a lot of the logic around handling the destination makes me wonder if things might be cleaner if we had websocket and datachannel transport classes which handled their logic and a transport manager or something that handled tracking which one was active or something.

Comment (bbaldino) [inline src/main/java/org/jitsi/videobridge/EndpointMessageTransport.java]: or maybe it's at least worth adding a helper to get the active connection so all that logic is tucked away somewhere else

Comment (bbaldino) [inline src/main/java/org/jitsi/videobridge/EndpointMessageTransport.java]: it seems like there's enough logic around which connection to use for a given message to warrant some sort of 'manager' there.

EDIT: have finished going through it all now...i guess it isn't too bad, but maybe a nice-to-have.  what do you think?

File: src/main/java/org/jitsi/videobridge/Conference.java (hunk 2 above)
Comment (bgrozev) [inline src/main/java/org/jitsi/videobridge/Conference.java]: The TransportManager needs to advertise the URL which endpoints can connect to. The URL encodes the conference and endpoint ID. And channelBundleId coincides with the endpoint ID.

File: src/main/java/org/jitsi/videobridge/EndpointMessageTransport.java (hunk 3 above)
Comment (bgrozev) [inline src/main/java/org/jitsi/videobridge/EndpointMessageTransport.java]: Yes, will do.

File: doc/rest.md (hunk 1 above)
Comment (bgrozev) [inline doc/rest.md]: Fixed.

File: src/main/java/org/jitsi/videobridge/EndpointMessageTransport.java (hunk 4 above)
Comment (bgrozev) [inline src/main/java/org/jitsi/videobridge/EndpointMessageTransport.java]: Good point, I'll add a helper.

File: src/main/java/org/jitsi/videobridge/Conference.java (hunk 2 above)
Comment (bbaldino) [inline src/main/java/org/jitsi/videobridge/Conference.java]: 👍

**Pull request merged**